### ISSUE TITLE
Azure dynamic host catalog returns nothing 

### ISSUE THREAD

**Issue by jchenship** [hashicorp/boundary#2348]
**Describe the bug**
Please see below for the steps

**To Reproduce**

Please also see my boundary dev server logs https://gist.github.com/jchenship/3115e71db0439d55f4b702cac1a3e044

If I read it correctly, there's no log related why it returns no hosts from azure dynamic host catalog (no error, no log saying the query result is empty etc)

I use the same steps as in the tutorial 

```
$ boundary version

Version information:
  Git Revision:        15f8f63b2d11d750e03eb5d2318edf1565560d93
  Version Number:      0.9.1
```

```
$ boundary dev > boundary.log 2>&1
```
```
$ boundary authenticate password -auth-method-id ampw_1234567890 -login-name admin
Password is not set as flag or in env, please enter it now (will be hidden):

Authentication information:
  Account ID:      acctpw_1234567890
  Auth Method ID:  ampw_1234567890
  Expiration Time: Thu, 18 Aug 2022 18:57:43 AEST
  User ID:         u_1234567890
Error opening "pass" keyring: Specified keyring backend not available
The token was not successfully saved to a system keyring. The token is:

at_hsrcV0HBY9_s12mcoLV51jyU7PojuK5jwdoX6DMTFU2Sum8DANzX8vPUsM6mNtKS7tj5bWVdxfqLXsn1a6qk22GLEehKF79fewEQxz5DqR6Cdo1n2B6NTp345ucvYc3jbhAqrgxCE6Pb4GuhJcki53f

It must be manually passed in via the BOUNDARY_TOKEN env var or -token flag. Storing the token can also be disabled via -keyring-type=none.
```
```
$ export BOUNDARY_TOKEN=at_hsrcV0HBY9_s12mcoLV51jyU7PojuK5jwdoX6DMTFU2Sum8DANzX8vPUsM6mNtKS7tj5bWVdxfqLXsn1a6qk22GLEehKF79fewEQxz5DqR6Cdo1n2B6NTp345ucvYc3jbhAqrgxCE6Pb4GuhJcki53f
```
```
$ boundary host-catalogs create plugin \                      
  -scope-id p_1234567890 \
  -plugin-name azure \
  -attr disable_credential_rotation=true \
  -attr tenant_id=env://ARM_TENANT_ID \
  -attr subscription_id=env://ARM_SUBSCRIPTION_ID \
  -attr client_id=env://ARM_CLIENT_ID \
  -secret secret_value=env://ARM_CLIENT_SECRET

Host Catalog information:
  Created Time:                  Thu, 11 Aug 2022 18:58:35 AEST
  ID:                            hc_O7qylzaWh2
  Plugin ID:                     pl_v1IRPorNHd
  Secrets HMAC:                  3pQAGGi4LyZTMvEMfTXr6N2jC26uUe5JheoRSNSGUDFQ
  Type:                          plugin
  Updated Time:                  Thu, 11 Aug 2022 18:58:35 AEST
  Version:                       1

  Scope:
    ID:                          p_1234567890
    Name:                        Generated project scope
    Parent Scope ID:             o_1234567890
    Type:                        project

  Plugin:
    ID:                          pl_v1IRPorNHd
    Name:                        azure

  Attributes:
    client_id:                   [REDACTED]
    disable_credential_rotation: true
    subscription_id:             [REDACTED]
    tenant_id:                   [REDACTED]

  Authorized Actions:
    no-op
    read
    update
    delete

  Authorized Actions on Host Catalog's Collections:
    host-sets:
      create
      list
    hosts:
      list
```
```
$ export HOST_CATALOG_ID=hc_O7qylzaWh2                        
```
```
$ boundary host-sets create plugin \                          
  -name database \
  -host-catalog-id $HOST_CATALOG_ID \
  -attr filter="tagName eq 'env' and tagValue eq 'stg'"

Host Set information:
  Created Time:        Thu, 11 Aug 2022 18:59:11 AEST
  Host Catalog ID:     hc_O7qylzaWh2
  ID:                  hs_TTzbJ3Nzap
  Name:                database
  Type:                plugin
  Updated Time:        Thu, 11 Aug 2022 18:59:11 AEST
  Version:             1

  Scope:
    ID:                p_1234567890
    Name:              Generated project scope
    Parent Scope ID:   o_1234567890
    Type:              project

  Plugin:
    ID:                pl_v1IRPorNHd
    Name:              azure

  Attributes:
    filter:            tagName eq 'env' and tagValue eq 'stg'

  Authorized Actions:
    no-op
    read
    update
    delete
```
```
$ boundary hosts list -host-catalog-id $HOST_CATALOG_ID       
No hosts found
```

after 5 minutes 

```
$ boundary hosts list -host-catalog-id $HOST_CATALOG_ID       
No hosts found
```

I can verify my azure credential is correct via using az cli 

```
$ az login --service-principal -u $ARM_CLIENT_ID -p $ARM_CLIENT_SECRET --tenant $ARM_TENANT_ID
[
  {
    "cloudName": "AzureCloud",
    "homeTenantId": "[redacted]",
    "id": "[REDACTED]",
    "isDefault": true,
    "managedByTenants": [],
    "name": "sub-stg-tools-1",
    "state": "Enabled",
    "tenantId": "[redacted]",
    "user": {
      "name": "[redacted]",
      "type": "servicePrincipal"
    }
  }
]
```
```
$ az role assignment list --assignee $ARM_CLIENT_ID
[
  {
    "canDelegate": null,
    "condition": null,
    "conditionVersion": null,
    "description": "",
    "id": "[REDACTED]",
    "name": "[REDACTED]",
    "principalId": "[REDACTED]",
    "principalName": "[REDACTED]",
    "principalType": "ServicePrincipal",
    "roleDefinitionId": "[REDACTED]",
    "roleDefinitionName": "Reader",
    "scope": "/subscriptions/[REDACTED]",
    "type": "Microsoft.Authorization/roleAssignments"
  }
]
```
```
az role assignment list --assignee $ARM_CLIENT_ID | grep scope | grep $ARM_SUBSCRIPTION_ID
    "scope": "/subscriptions/[REDACTED]",
```
```
$ az resource list --tag env=stg --output table | grep -i virtualmachine
aks-general-24046850-vmss                                                 rg-tools-stg-kubernetes-aks-main  southeastasia  microsoft.compute/virtualmachinescalesets
aks-gitlabrunner-27009483-vmss                                            rg-tools-stg-kubernetes-aks-main  southeastasia  microsoft.compute/virtualmachinescalesets
aks-system-40813308-vmss                                                  rg-tools-stg-kubernetes-aks-main  southeastasia  microsoft.compute/virtualmachinescalesets
vm-tools-stg-network-bastion                                              rg-tools-stg-network              southeastasia  microsoft.compute/virtualmachines
vmext-tools-stg-network-bastion-aadssh                                    rg-tools-stg-network              southeastasia  microsoft.compute/virtualmachines/extensions
```

**Expected behavior**

**Additional context**

**Comment by jchenship** [comment]
also tried boundary 0.10.0 still nothing

**Comment by jefferai** [comment]
Hi @jchenship !

I just tried this with my own VMs and a similar filter and I get results:

```
Host Set information:
  Created Time:        Thu, 11 Aug 2022 12:48:09 EDT
  Host Catalog ID:     hcplg_2UstTz0UCs
  ID:                  hsplg_lWT7CtUnAU
  Type:                plugin
  Updated Time:        Thu, 11 Aug 2022 13:23:22 EDT
  Version:             24

  Scope:
    ID:                p_1234567890
    Name:              Generated project scope
    Parent Scope ID:   o_1234567890
    Type:              project

  Plugin:
    ID:                pl_7vxNsTWgfQ
    Name:              azure

  Attributes:
    filter:            tagName eq 'application' and tagValue eq 'south-seas'

  Authorized Actions:
    no-op
    read
    update
    delete

  Host IDs:
    hplg_C5w71fCHBP
    hplg_GVBnnzjlIX
```

If you're not seeing errors in Boundary's observations stream (which you should see when updating the host set with a filter if e.g. attributes or secrets are invalid) then there are a couple of checks in the code that maybe are tripping things up.

The first is that there is a validation that it's a compute vm, e.g. the type is `Microsoft.Compute/virtualMachines`. I don't know if Azure has any other kind of virtual machine that has a different qualifier.

The other is that we also range through the statues returned by the instance view on the VM and check that the powerstate value is `powerstate/running`. Since the available hosts are being updated dynamically if a host goes down we want to remove it as an endpoint that a target can allow a connection to, so we ensure that we are only populating running hosts.

Can you verify that your hosts are the same type and are in running state?

**Comment by jchenship** [comment]
Hi @jefferai , thank you for the quick response, I tried using the service principal to call resources api and vm instance view api, I think there might be some case mismatching 

the [code](https://github.com/hashicorp/boundary-plugin-host-azure/blob/main/plugin/plugin.go#L258) is filtering on `Microsoft.Compute/virtualMachines` while the api returns `microsoft.compute/virtualmachines` for me

```
$ az resource list --tag env=stg --output table | grep -ie 'Microsoft.Compute/virtualMachines\>'
vm-tools-stg-logging-sftp                                                 rg-tools-stg-logging              southeastasia  microsoft.compute/virtualmachines
vm-tools-stg-network-bastion                                              rg-tools-stg-network              southeastasia  microsoft.compute/virtualmachines
vm-tools-stg-openvpn-contractor                                           rg-tools-stg-openvpn              southeastasia  microsoft.compute/virtualmachines
vm-tools-stg-openvpn-main                                                 rg-tools-stg-openvpn              southeastasia  microsoft.compute/virtualmachines
```

```
$ az resource list --tag env=stg | grep -i '"Microsoft.Compute/virtualMachines"'
    "type": "microsoft.compute/virtualmachines"
    "type": "microsoft.compute/virtualmachines"
    "type": "microsoft.compute/virtualmachines"
    "type": "microsoft.compute/virtualmachines"
```

and this one the [code](https://github.com/hashicorp/boundary-plugin-host-azure/blob/main/plugin/plugin.go#L302) is expecting lowercase `prefix := "powerstate/"` but my API returns PascalCase

```
az vm get-instance-view -n vm-tools-stg-network-bastion -g rg-tools-stg-network  | grep -i powerstate
        "code": "PowerState/running",
```

**Comment by jefferai** [comment]
I think we certainly can do a case insensitive compare. Those cases were used based on the actual results we saw and they work for me but certainly something we can do. I'm skeptical it will make the difference here unfortunately as I think the most likely case is that the other client is doing formatting of its own rather than the api returning different values. But happy to make that change!

**Comment by jchenship** [comment]
actually the powerstate comparison is already normalized so it shouldn't be powerstate https://github.com/hashicorp/boundary-plugin-host-azure/blob/main/plugin/plugin.go#L301, but it could still be the resource type comparison.

> the other client is doing formatting of its own rather than the api returning different values. 

I used `--debug` flag can checked my api json response is actually returning all lowercase, unless the api raw response of `--debug` flag is also formatted.  This sounds weird to me as well, but it does seem like a typical Microsoft thing .

**Comment by jefferai** [comment]
I have a draft PR at https://github.com/hashicorp/boundary-plugin-host-azure/pull/8

There are a few people internally that can replicate your issue so I'll work with them to try to see if that fixes it, or dig into what's going on in their accounts.

**Comment by jefferai** [comment]
Hi @jchenship 

Not good news at the moment: not only can nobody around here, including myself, make tag-based filters work at the moment, but my own test from six days ago (https://github.com/hashicorp/boundary/issues/2348#issuecomment-1212280521) is also not working for me any longer.

@stellarsquall went back to a version of Boundary from when he wrote his learn guide and no luck there either. So I have no idea what the difference could be between us using the Go libs and whatever the `az` CLI is doing but something is crazily flaky.

I did notice in digging into this that they have deprecated the SDK package we are using so I'm going to jump to their new package in case it makes some kind of difference. Otherwise I will start trying to engage more resources internally to try to get either more knowledgeable people here or to reach our through our partnership with Microsoft to try to get some clarity, because when our code hasn't changed at all and it's either not working or working intermittently then it feels like this is some issue external to Boundary.

**Comment by jefferai** [comment]
Oh, and the moment I type that^, suddenly tag filters are working fine for me...

**Comment by jefferai** [comment]
I've deleted my earlier comments because I was wrong.

When I started updating this branch I updated various deps and I was doing it on a different branch from the case-insensitivity work in that PR ^. I did that specifically because I wanted to dig into this without side effects from that PR, since the default filter (`resourceType eq 'Microsoft.Compute/virtualMachines'`) was working and returning items.

I now know that case-insensitivity PR was actually the reason it worked. I still don't know what, or why, things may have changed on the Azure side -- and I'm convinced, based on @stellarsquall going back and trying with older Boundary versions, and the fact that we didn't change our deps, that something did change _on the Azure side_. When I look at the returned values I do see that, yes, the type is coming back lowercase whereas before it would _always_ come back mixed-case, and in fact when you use the default filter it also comes back mixed case.

I mean, this is kinda nuts. Check out what's returned with the default filter I put above, on resourceType:

```
2022/08/17 11:56:03 name: boundary-vm1-dev type: Microsoft.Compute/virtualMachines
2022/08/17 11:56:03 name: boundary-vm2-dev type: Microsoft.Compute/virtualMachines
```

Now check it out with a tag filter of `tagName eq 'application' and tagValue eq 'dev'`

```
2022/08/17 11:58:43 name: boundary-vm1-dev type: microsoft.compute/virtualmachines
2022/08/17 11:58:43 name: boundary-vm1-dev_OsDisk_1_5c19204db333482ea382d976520b520b type: microsoft.compute/disks
2022/08/17 11:58:43 name: boundary-vm2-dev type: microsoft.compute/virtualmachines
```

It literally changes case coming back depending on what filter you use.

So, that case insensitivity patch is definitely a part of the fix (and thanks for catching that!). This really did used to work. And the current behavior is hard to grok, but it is what it is. I also found in the course of the investigation that they have deprecated one of the Azure sdk packages we're using so I'm going to update that as well. Sit tight but probably this should be out and fixed in 0.10.3.
